marshmallow/fields.py

```diff
@@ -1355,6 +1350,45 @@ def _deserialize(self, value, *args, **kwargs):
         return self.constant
 
 
+class Default(Field):
+    """A field that infers how to serialize, based on the value type."""
+
+    TYPE_MAPPING = {
+        text_type: String,
+        binary_type: String,
+        dt.datetime: DateTime,
+        float: Float,
+        bool: Boolean,
+        tuple: Raw,
+        list: Raw,
+        set: Raw,
+        int: Integer,
+        uuid.UUID: UUID,
+        dt.time: Time,
+        dt.date: Date,
+        dt.timedelta: TimeDelta,
+        decimal.Decimal: Decimal,
+    }
+
+    def __init__(self):
+        super(Default, self).__init__()
+
+        self._field = None
+
+    def _serialize(self, value, attr, obj):
+        field_type = self.TYPE_MAPPING.get(type(value), Field)
+        if field_type is Field:
+            return super(Default, self)._serialize(value, attr, obj)
+
+        field = self._field
+        if type(field) is not field_type:
+            self._field = field = field_type()
+            field._bind_to_schema(self.name, self.parent)
+
+        # Use the local field rather than self._field for thread safety.
+        return field._serialize(value, attr, obj)
```

Comment:
I don't understand the need for `self._field`. Is it just some sort of optimization: avoid binding to schema at every call to deserialize, caching the instantiated and bound field instead? If so, is it worth doing? Would it be worth going further and rather than caching only the field for the last type, use a `dict` to cache a field instance for each new type that is encountered?

Please forgive me if I am totally mistaken. I may have missed the obvious.

Reply:
OK, what I thought but I wasn't sure.

You've been investigating performance issues much more than I did and I have no idea of the penalty induced by instantiation.

The `dict` could also be filled at import time to avoid any penalty when serializing.

Reply:
What do you mean? The fields here are bound to the schema, so they can't be shared across schemas.

Reply:
Oh, I don't have performance numbers or anything, it just seemed like having a single-field memo would give us what we wanted in most cases, while not adding more than a few lines of code.

Reply:
I don't really mind either way.

When reviewing your code, I tried to understand (see my first question) and found it surprising. I think the implementation I propose is marginally superior from a functional point of view as it covers that corner case when the type would change, with no extra performance cost (or am I wrong?) and I find it easier to grasp at first glance, but that's just my POV.

Anyway, I don't want to nitpick. This is just about optimizing or not a corner case. What matters the most to me is to keep the code simple and easy to understand.

I'm fine with whatever you and the other reviewer(s) find best.

Reply:
Your example is actually not entirely safe against certain edge cases, incidentally.

In the event where the user dynamically generates new classes on-the-fly, the cache could potentially grow without bound and lead to excess memory usage. The correct caching key needs to be either the _field_ class, or else the cache needs to be a weakmap.

Granted, this is extremely unlikely to cause issues in practice, but I do think it's a point in favor of going for a simpler cache here.

Reply:
Good catch.

Here's an alternative implementation that should address this (using the field class as key).

```py
    def __init__(self):
        super(Inferred, self).__init__()
        self._field_cache = {}

    def _serialize(self, value, attr, obj):
        field_cls = self.TYPE_MAPPING.get(type(value), None)
        if field_cls is None:
            field = super(Inferred, self)
        else:
            field = self._field_cache.get(field_cls, None)
            if field is None:
                field = field_cls()
                field._bind_to_schema(self.name, self.parent)
                self._field_cache[field_cls] = field

        # Use the local field rather than self._field for thread safety.
        return field._serialize(value, attr, obj)
```

Other minor differences with your implementation:

- use `None` instead of unused `Field`
- `_serialize(value, attr, obj)` appears only once at the end of the method

Reply:
Benchmark: https://gist.github.com/deckar01/7643bb78a2dd04375fd2c16f90b9f052

10k iterations with python 3.7.0 dumping 3 inferred fields:

| (s) | dev | revamp-infer-fields | lafrech-cache |
|---|---|---|---|
| **static** | 1.01s | 0.316s | 0.331s |
| **changing** | 1.03s | 0.908s | 0.329s |

| (%) | dev | revamp-infer-fields | lafrech-cache |
|---|---|---|---|
| **static** | 100% | 31.3% | 32.8% |
| **changing** | 100% | 88.2% | 31.9% |

The speed up for dumping inferred fields is significant, so the optimization seems to justify the added complexity. Since the type cache @lafrech proposed optimizes both use cases without increasing the code complexity or penalizing the primary use case, it would advocate using it.

Reply:
I'm okay with either if you all feel strongly about this. At the margins I look at the non-dict based solution as being slightly simpler, but I'd be okay with using something dict-based as well.